marketing/tasks.py

```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015-2020 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Worker tasks for marketing features
+"""
+
+# Imports #####################################################################
+
+import logging
+
+from django.conf import settings
+from django.template import Context
+from django.utils import timezone
+from huey import crontab
+from huey.contrib.djhuey import db_periodic_task
+
+from marketing.models import Subscriber, EmailTemplate, SentEmail
+from registration.models import BetaTestApplication
+
+
+# Logging #####################################################################
+
+logger = logging.getLogger(__name__)
+
+
+# Functions ###################################################################
+
+def render_email_and_send(template: EmailTemplate, application: BetaTestApplication):
+    """
+    """
+    context = Context({
+        "full_name": application.user.get_full_name(),
+        "username": application.user.get_username(),
+        "instance_name": application.instance.name,
+        "subdomain": application.subdomain,
+        "application": application
+    })
+    subject = template.subject_template.render(context)
+    html_body = template.html_body_template.render(context)
+    plaintext_body = template.plaintext_body_template.render(context)
+
+    if not SentEmail.objects.filter(recipient_email=application.user.email, template=template).exists()
+        send_mail(
+            subject=subject,
+            message=plaintext_body,
+            html_message=html_body,
+            from_email=settings.DEFAULT_FROM_EMAIL,
+            recipient_list=[application.user.email],
+        )
+        SentEmail.objects.create(
+            recipient_email=application.user.email,
+            template=template,
+            subject=subject,
+            html_body=html_body,
+            plaintext_body=plaintext_body,
+        )
+
+
+# Tasks #######################################################################
+
+@db_periodic_task(crontab(day='*', hour='1', minute='0'))
+def send_followup_emails():
```

Comment:
The logic between this and `render_email_and_send` doesn't have a clear division of responsibilities, and it also doesn't look like it'll scale well. You can make use of `prefetch` to avoid extra queries here.